### PR TITLE
detect incompatible CI LLVM options more precisely

### LINKED ISSUE #129153: improve incompatible option detection for ci-llvm
Since #128822, CI builders provide their configuration in the `builder-config` file, which can be used to more precisely detect incompatible ci-llvm options (similar to how it's done for rustc options in https://github.com/rust-lang/rust/pull/129052). This helps to prevent incorrect assumptions of CI builders on a used llvm option which was also used while building ci-llvm artifact.

### PR DESCRIPTION
**Pull request by onur-ozkan**
Previously, the logic here was simply checking whether the option was set in `config.toml`. This approach was not manageable in our CI runners as we set so many options in config.toml. In reality, those values are not incompatible since they are usually the same value used to generate the CI llvm. Now, the new logic compares the configuration values with the values used to generate the CI llvm, so we get more precise results and make the process more manageable.

Fixes #129153

### CODE REVIEW

File: src/bootstrap/src/core/download.rs [hunk 1]
```diff
@@ -700,17 +703,43 @@ download-rustc = false
...
+                let current_config_toml = Self::get_toml(config_path).unwrap();
+
+                match self.get_builder_toml("ci-llvm") {
+                    Ok(ci_config_toml) => {
+                        check_incompatible_options_for_ci_llvm(current_config_toml, ci_config_toml)
+                            .unwrap();
+                    }
+                    Err(e) if e.to_string().contains("unknown field") => {
+                        println!(
+                            "WARNING: CI rustc has some fields that are no longer supported in bootstrap; download-rustc will be disabled."
```
Comment (Kobzol) [inline src/bootstrap/src/core/download.rs]: `download-rustc` -> `download-ci-llvm`

File: src/bootstrap/src/core/config/config.rs [hunk 2]
```diff
@@ -2823,6 +2820,7 @@ impl Config {
 
 /// Compares the current `Llvm` options against those in the CI LLVM builder and detects any incompatible options.
 /// It does this by destructuring the `Llvm` instance to make sure every `Llvm` field is covered and not missing.
+#[cfg(not(test))]
```
Comment (Kobzol) [inline src/bootstrap/src/core/config/config.rs:2827]: Why is this here? (and why doesn't it use `#[cfg(not(feature = "bootstrap-self-test"))]` instead? :thinking:).

**Review by Kobzol — COMMENTED** [review]
Ooooh, I have wanted this for so long! Great job :heart:

Left two nits.

File: src/bootstrap/src/core/config/config.rs (hunk 2 above)
Comment (onur-ozkan) [reply src/bootstrap/src/core/config/config.rs]: Both works actually. Changed it to "bootstrap-self-test".

**Comment by Kobzol** [comment]
Thank you!

@bors r+ rollup=never

**Comment by onur-ozkan** [comment]
Rebased.

@bors r=Kobzol
